frontend/src/pages/add/VendorAdd.tsx

```diff
@@ -0,0 +1,212 @@
+import { useRef, useState } from "react";
+import { Button } from "primereact/button";
+import { logger } from "../../util/Logger";
+import { Toast } from "primereact/toast";
+import { useNavigate } from "react-router-dom";
+import { showFailure, showSuccess } from "../../components/Toast";
+import { DataTable } from "primereact/datatable";
+import { Column } from "primereact/column";
+import { TableColumn, createColumns } from "../../components/TableColumns";
+import { percentEditor, textEditor } from "../../util/TableCellEditFuncs";
+import { findById } from "../../util/IDOperations";
+import { useImmer } from "use-immer";
+import React from "react";
+import { v4 as uuid } from "uuid";
+import ConfirmPopup from "../../components/popups/ConfirmPopup";
+import { Toolbar } from "primereact/toolbar";
+import { VENDORS_API, AddVendorReq } from "../../apis/VendorsAPI";
+import axios from "axios";
+import AddRowButton from "../../components/buttons/AddRowButton";
+import BackButton from "../../components/buttons/BackButton";
+
+export interface VendorRow {
+  id: string;
+  vendorName: string;
+  buybackRate: number | undefined;
+}
+
+export default function GenreAdd() {
+  const emptyVendor: VendorRow = {
+    id: uuid(),
+    vendorName: "",
+    buybackRate: undefined,
+  };
+
+  const [vendors, setVendors] = useImmer<VendorRow[]>([]);
+  const [lineData, setLineData] = useState<VendorRow>(emptyVendor);
+  const [isConfirmationPopupVisible, setIsConfirmationPopupVisible] =
+    useState<boolean>(false);
+  const [isGoBackActive, setIsGoBackActive] = useState<boolean>(false);
+
+  const COLUMNS: TableColumn[] = [
+    {
+      field: "vendorName",
+      header: "Vendor Name",
+      customBody: (rowData: VendorRow) =>
+        textEditor(rowData.vendorName, (newValue) => {
+          setVendors((draft) => {
+            const vendor = findById(draft, rowData.id);
+            vendor!.vendorName = newValue;
```

Comment:
Not a big deal, but I realized myself that you can add a ! after findById(...)! and then just use vendor.vendorName after that. Doesn't matter in this case since but in the future it saves you from having to write a ! a bunch of times

Reply:
noted